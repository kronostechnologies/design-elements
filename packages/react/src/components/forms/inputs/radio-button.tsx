import React, { ChangeEvent, Fragment, ReactElement } from 'react';
import styled from 'styled-components';

const StyledLabel = styled.label `
    ${(props: {disabled?: boolean}) => props.disabled ? null : 'cursor: pointer;'}
    display: block;
    font-size: 0.875rem;
    padding-left: 25px;
    position: relative;
    margin: 15px 0 0 0;
    user-select: none;
    input {
        display: none;
        &:checked + .circle {
            background-color: #0080a5;
            border: 1px solid #0080a5;
            &:after {
                background-color: white;
                border-radius: 50%;
                content: "";
                height: 7px;
                left: 50%;
                position: absolute;
                top: 50%;
                transform: translate(-50%, -50%);
                width: 7px;
            }
        }
    }
    .circle {
        align-self: center;
        background-color: ${(props: {disabled?: boolean}) => props.disabled ? '#f1f2f2' : '#ffffff'};
        border: 1px solid ${(props: {disabled?: boolean}) => props.disabled ? '#d9dde2' : '#57666e'};
        border-radius: 50%;
        display: inline-block;
        height: 16px;
        left: 0;
        margin-top: 2px;
        position: absolute;
        top: 0;
        width: 16px;
    }
    &:hover .circle {
        border: 1px solid ${(props: {disabled?: boolean}) => props.disabled ? '#d9dde2' : '#0080a5'};
    }
`;

interface RadioButtonProps {
    label?: string;
    /** Sets the name property of all buttons */
    groupName: string;
    buttons: {
        label: string,
        value: string,
        defaultChecked?: boolean,
        disabled?: boolean,
    }[];
    onChange?(event: ChangeEvent<HTMLInputElement>): void;
}

export function RadioButton({ buttons, groupName, label, onChange }: RadioButtonProps): ReactElement  {
    const handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
        if (onChange) {
            onChange(event);
        }
    };

    return (
        <Fragment>
            <label style={{ fontSize: '0.75rem' }}>{label}</label>
            {buttons.map((button, key) => (
                <StyledLabel
                    disabled={button.disabled}
                    key={groupName + '-' + key}
                    style={key === 0 ? { marginTop: '7px' } : {}}
                > {button.label}
                    <input
                        type="radio"
                        name={groupName}
                        value={button.value}
                        defaultChecked={button.defaultChecked}
                        disabled={button.disabled}
                        onChange={handleChange}
                    />
                    <span className="circle" />
                </StyledLabel>
            ))}
        </Fragment>
    );
}
