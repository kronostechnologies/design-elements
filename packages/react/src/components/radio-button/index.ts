export { RadioButton } from './radio-button';
export type { RadioButtonProps } from './radio-button';
