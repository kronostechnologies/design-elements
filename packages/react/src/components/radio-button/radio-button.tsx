import { ChangeEvent, FunctionComponent } from 'react';
import styled from 'styled-components';
import { focus } from '../../utils/css-state';
import { useId } from '../../hooks/use-id';

const getDotSvgDataUrl = (color: string): string => {
    const svg = `
        <svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
            <circle cx="8" cy="8" r="4" fill="${color}"/>
        </svg>`;
    return `url('data:image/svg+xml;utf8,${encodeURIComponent(svg)}')`;
};

const StyledInput = styled.input<{ disabled?: boolean }>`
    appearance: none;
    background-color: ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-background-color'] : theme.component['radio-button-background-color'])};
    border: 1px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-border-color'] : theme.component['radio-button-border-color'])};
    border-radius: 50%;
    color: ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-background-color'] : theme.component['radio-button-background-color'])};
    display: inline-block;
    flex-shrink: 0;
    height: var(--size-1x);
    margin: 0;
    margin-top: var(--spacing-half);
    position: relative;
    width: var(--size-1x);
    
    ${(theme) => focus(theme, { selector: '+' })}
     
    &:checked {
        background-image: ${({ theme, disabled }) => getDotSvgDataUrl(disabled ? theme.component['radio-button-disabled-checked-dot-color'] : theme.component['radio-button-checked-dot-color'])};
        background-position: center;
        background-repeat: no-repeat;
        border: 2px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-border-color'] : theme.component['radio-button-checked-border-color'])};
    }
    
    &:disabled {
        & + label {
            color: ${({ theme }) => theme.component['radio-button-disabled-label-color']};
        }
    }
        
    &:hover {
        &:checked:not(:disabled) {
            border: 1px solid ${({ theme }) => theme.component['radio-button-hover-border-color']};
        }
        
        &:not(:checked) {
            border: 1px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-hover-border-color'] : theme.component['radio-button-hover-border-color'])};
        }
    }
`;

const StyledLabel = styled.label`
    font-size: 0.875rem;
    line-height: 1.5rem;
    margin-left: var(--spacing-1x);
`;

const StyledContainer = styled.div`
    align-items: flex-start;
    display: inline-flex;
    margin-top: var(--spacing-1x);
    position: relative;
`;

interface RadioButtonProps {
    ariaLabel?: string;
    ariaLabelledBy?: string[];
    checked?: boolean;
    className?: string;
    defaultChecked?: boolean;
    disabled?: boolean;
    id?: string;
    label?: string;
    name?: string;
    value?: string;

    onChange?(event: ChangeEvent<HTMLInputElement>): void;
}

export const RadioButton: FunctionComponent<RadioButtonProps> = ({
    ariaLabel,
    ariaLabelledBy,
    checked,
    className,
    defaultChecked,
    disabled,
    id,
    label,
    name,
    value,
    onChange,
}) => {
    const inputId = useId(id);

    return (
        <StyledContainer>
            <StyledInput
                data-testid={`radiobutton-${inputId}`}
                id={inputId}
                type="radio"
                name={name}
                className={className}
                value={value}
                defaultChecked={defaultChecked}
                checked={checked}
                disabled={disabled}
                onChange={onChange}
                aria-label={ariaLabel}
                aria-labelledby={ariaLabelledBy?.join(' ')}
            />
            {
                label && (
                    <StyledLabel
                        data-testid={`radiobutton-${inputId}_label`}
                        htmlFor={inputId}
                    >
                        {label}
                    </StyledLabel>
                )
            }
        </StyledContainer>
    );
};

RadioButton.displayName = 'RadioButton';
