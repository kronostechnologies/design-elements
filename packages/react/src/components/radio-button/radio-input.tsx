import { ChangeEvent, FocusEvent, forwardRef, Ref } from 'react';
import styled from 'styled-components';
import { useId } from '../../hooks/use-id';
import { focus } from '../../utils/css-state';

const getDotSvgDataUrl = (color: string): string => {
    const svg = `
        <svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
            <circle cx="8" cy="8" r="4" fill="${color}"/>
        </svg>`;
    return `url('data:image/svg+xml;utf8,${encodeURIComponent(svg)}')`;
};

const StyledInput = styled.input<{ disabled?: boolean; isMobile?: boolean }>`
    appearance: none;
    background-color: ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-background-color'] : theme.component['radio-button-background-color'])};
    border: 1px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-border-color'] : theme.component['radio-button-border-color'])};
    border-radius: 50%;
    color: ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-background-color'] : theme.component['radio-button-background-color'])};
    display: inline-block;
    flex-shrink: 0;
    height: var(--size-1x);
    margin: 0;
    position: relative;
    width: var(--size-1x);

    ${focus}

    &:checked {
        background-image: ${({ theme, disabled }) => getDotSvgDataUrl(disabled ? theme.component['radio-button-disabled-checked-dot-color'] : theme.component['radio-button-checked-dot-color'])};
        background-position: center;
        background-repeat: no-repeat;
        border: 2px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-border-color'] : theme.component['radio-button-checked-border-color'])};
    }

    &:disabled {
        & + label {
            color: ${({ theme }) => theme.component['radio-button-disabled-label-color']};
        }
    }

    &:hover {
        &:checked:not(:disabled) {
            border: 1px solid ${({ theme }) => theme.component['radio-button-hover-border-color']};
        }

        &:not(:checked) {
            border: 1px solid ${({ theme, disabled }) => (disabled ? theme.component['radio-button-disabled-hover-border-color'] : theme.component['radio-button-hover-border-color'])};
        }
    }
`;

export interface RadioInputProps {
    ariaLabel?: string;
    ariaLabelledBy?: string[];
    checked?: boolean;
    className?: string;
    defaultChecked?: boolean;
    disabled?: boolean;
    id?: string;
    name?: string;
    required?: boolean;
    value?: string;

    onBlur?(event: FocusEvent<HTMLInputElement>): void;
    onChange?(event: ChangeEvent<HTMLInputElement>): void;
}

export const RadioInput = forwardRef<HTMLInputElement, RadioInputProps>(({
    ariaLabel,
    ariaLabelledBy,
    checked,
    className,
    defaultChecked,
    disabled,
    id,
    name,
    value,
    required,
    onBlur,
    onChange,
}: RadioInputProps, ref: Ref<HTMLInputElement>) => {
    const inputId = useId(id);

    return (
        <StyledInput
            data-testid={`radiobutton-${inputId}`}
            id={inputId}
            type="radio"
            name={name}
            className={className}
            value={value}
            defaultChecked={defaultChecked}
            checked={checked}
            disabled={disabled}
            onBlur={onBlur}
            onChange={onChange}
            required={required}
            aria-label={ariaLabel}
            aria-labelledby={ariaLabelledBy?.join(' ')}
            ref={ref}
        />
    );
});

RadioInput.displayName = 'RadioInput';
