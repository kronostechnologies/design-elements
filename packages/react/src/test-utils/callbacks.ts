export function doNothing(): void {
    // do nothing
}
