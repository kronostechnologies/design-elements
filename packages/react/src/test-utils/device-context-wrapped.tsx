import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
