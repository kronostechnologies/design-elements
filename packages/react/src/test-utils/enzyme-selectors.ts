import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
