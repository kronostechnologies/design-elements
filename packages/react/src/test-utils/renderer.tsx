import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
