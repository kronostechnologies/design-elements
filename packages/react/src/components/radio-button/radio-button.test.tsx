import { mountWithTheme } from '../../test-utils/renderer';
import { RadioButton } from './radio-button';
import { getByTestId } from '../../test-utils/enzyme-selectors';
import { doNothing } from '../../test-utils/callbacks';

describe('Radio button', () => {
    test('should have controllable data-testid', () => {
        const wrapper = mountWithTheme(<RadioButton checked data-testid="radiobutton-testid" />);

        expect(getByTestId(wrapper, 'radiobutton-testid').exists()).toBe(true);
    });

    test('should be checked by default is defaultChecked prop is present', () => {
        const wrapper = mountWithTheme(<RadioButton defaultChecked />);

        const input = wrapper.find('input[type="radio"]');
        expect(input.prop('defaultChecked')).toBe(true);
    });

    test('onChange callback should be called when radio button is checked', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<RadioButton onChange={callback} />);

        wrapper.find('input[type="radio"]').simulate('change');

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('should be checked when checked prop is set to true', () => {
        const wrapper = mountWithTheme(<RadioButton checked onChange={doNothing} />);

        const input = wrapper.find('input[type="radio"]');
        expect(input.prop('checked')).toBe(true);
    });

    test('should be disabled when disabled prop is set to true', () => {
        const wrapper = mountWithTheme(<RadioButton disabled />);

        const input = wrapper.find('input[type="radio"]');
        expect(input.prop('disabled')).toBe(true);
    });

    test('matches snapshot', () => {
        const tree = mountWithTheme(<RadioButton label="This is a label" />);

        expect(tree).toMatchSnapshot();
    });
});
