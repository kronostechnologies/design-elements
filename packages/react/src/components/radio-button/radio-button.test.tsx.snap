// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio button matches snapshot 1`] = `
.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: 50%;
  color: #FFFFFF;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: var(--size-1x);
  margin: 0;
  margin-top: var(--spacing-half);
  position: relative;
  width: var(--size-1x);
}

.c1 + {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c1:focus + {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1:checked {
  border: 2px solid #006296;
}

.c1:checked::after {
  background-color: #006296;
  border-radius: 50%;
  content: '';
  display: block;
  height: var(--size-half);
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  width: var(--size-half);
}

.c1:hover {
  background-color: #DBDEE1;
  border: 1px solid #000000;
}

.c2 {
  font-size: 0.875rem;
  line-height: 1.5rem;
  margin-left: var(--spacing-1x);
}

.c0 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-top: var(--spacing-1x);
  position: relative;
}

<RadioButton
  label="This is a label"
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.input
        data-testid="radiobutton-uuid1"
        id="uuid1"
        type="radio"
      >
        <input
          className="c1"
          data-testid="radiobutton-uuid1"
          id="uuid1"
          type="radio"
        />
      </styled.input>
      <styled.label
        data-testid="radiobutton-uuid1_label"
        htmlFor="uuid1"
      >
        <label
          className="c2"
          data-testid="radiobutton-uuid1_label"
          htmlFor="uuid1"
        >
          This is a label
        </label>
      </styled.label>
    </div>
  </styled.div>
</RadioButton>
`;
